Lib/test/test_ctypes/test_dlerror.py

```diff
@@ -0,0 +1,97 @@
+import os
+import sys
+import unittest
+from ctypes import CDLL
+
+FOO_C = r"""
+#include <stdio.h>
+
+/* This is a 'GNU indirect function' (IFUNC) that will be called by
+   dlsym() to resolve the symbol "foo" to an address. Typically, such
+   a function would return the address of an actual function, but it
+   can also just return NULL.  For some background on IFUNCs, see
+   https://willnewton.name/uncategorized/using-gnu-indirect-functions.
+
+   Adapted from Michael Kerrisk's answer: https://stackoverflow.com/a/53590014.
+*/
+
+asm (".type foo, @gnu_indirect_function");
+
+void *foo(void)
+{
+    fprintf(stderr, "foo IFUNC called\n");
+    return NULL;
+}
+"""
+
+
+@unittest.skipUnless(sys.platform.startswith('linux'),
+                     'Test only valid for Linux')
+class TestNullDlsym(unittest.TestCase):
+    """GH-126554: Ensure that we catch NULL dlsym return values
+
+    In rare cases, such as when using GNU IFUNCs, dlsym(),
+    the C function that ctypes' CDLL uses to get the address
+    of symbols, can return NULL.
+
+    The objective way of telling if an error during symbol
+    lookup happened is to call glibc's dlerror() and check
+    for a non-NULL return value.
+
+    However, there can be cases where dlsym() returns NULL
+    and dlerror() is also NULL, meaning that glibc did not
+    encounter any error.
+
+    In the case of ctypes, we subjectively treat that as
+    an error, and throw a relevant exception.
+
+    This test case ensures that we correctly enforce
+    this 'dlsym returned NULL -> throw Error' rule.
+    """
+
+    def test_null_dlsym(self):
+        import subprocess
+        import tempfile
+
+        # When using 'shell=True', Python invokes
+        # /bin/sh -c 'args[0]' args[1] ...
+        #
+        # The shell uses arg[0] as the
+        # command to execute,
+        # while the rest are arguments
+        # to the shell itself.
+        #
+        # To avoid mistakes, pass a single string
+        # (equivalent to a single-elem list),
+        # with the shell command to execute.
+        retcode = subprocess.call("gcc --version",
```

Comment:
```suggestion
        # Recall: using shell=True is equivalent to: /bin/sh -c 'args[0]' args[1] ...,
        # so we need to pass a list with a single item, namely the command to run.
        retcode = subprocess.call(["gcc --version"],
```

Since we are saying "args[0]" and "args[1]", it's better to keep the list-form. Without the comment, the string form could be fine but with the comment, the list form makes more sense (otherwise you need to mention that it's equivalent to a 1-element list).

Reply:
It's more concise this way. Thanks!